recipes/uni-algo/all/conanfile.py

```diff
@@ -1,66 +1,147 @@
+            if self.options.get_safe("shared"):
+                raise ConanInvalidConfiguration(f"{self.ref} can not be built as shared on Visual Studio and msvc.")
 
     def source(self):
         get(self, **self.conan_data["sources"][self.version], strip_root=True)
 
+    def generate(self):
+        tc = CMakeToolchain(self)
+        if self.options.header_only:
+            tc.variables["UNI_ALGO_HEADER_ONLY"] = True
+        elif self.options["shared"]:
+            tc.variables["BUILD_SHARED_LIBS"] = True
+
+        if is_msvc(self):
+            # don't use self.settings.compiler.runtime
+            tc.variables["USE_MSVC_RUNTIME_LIBRARY_DLL"] = not is_msvc_static_runtime(self)
+
+        tc.generate()
+
+    def build(self):
+        if not self.options.header_only:
+            cmake = CMake(self)
+            cmake.configure()
+            cmake.build()
+
     def package(self):
         copy(self, pattern="LICENSE.md", dst=os.path.join(self.package_folder, "licenses"), src=self.source_folder)
-        copy(
-            self,
-            pattern="*.h",
-            dst=os.path.join(self.package_folder, "include"),
-            src=os.path.join(self.source_folder, "include"),
-        )
+        if self.options.header_only:
+            copy(
+                self,
+                pattern="*.h",
+                dst=os.path.join(self.package_folder, "include"),
+                src=os.path.join(self.source_folder, "include"),
+            )
+        else:
+            cmake = CMake(self)
+            cmake.install()
+
+            rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
+            rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))
+            rmdir(self, os.path.join(self.package_folder, "share"))
+            rm(self, "*.la", os.path.join(self.package_folder, "lib"))
+            rm(self, "*.pdb", os.path.join(self.package_folder, "lib"))
+            rm(self, "*.pdb", os.path.join(self.package_folder, "bin"))
 
     def package_info(self):
-        self.cpp_info.bindirs = []
-        self.cpp_info.libdirs = []
+        if self.options.header_only:
+            self.cpp_info.bindirs = []
+            self.cpp_info.libdirs = []
+        else:
+            if self.settings.build_type == "Debug":
+                if Version(self.version) < "0.7":
+                    self.cpp_info.libs = [f"{self.name}d"]
+                else:
+                    self.cpp_info.libs = [f"{self.name}-debug"]
+            else:
+                self.cpp_info.libs = [f"{self.name}"]
+            if self.settings.os in ["Linux", "FreeBSD"]:
+                self.cpp_info.system_libs.append("m")
+                self.cpp_info.system_libs.append("pthread")
+                self.cpp_info.system_libs.append("dl")
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = f"{self.name}".upper()
```

Comment:
There's look like extras from the template, does upstream install cmake config files?

Reply:
I guess you mean the upstream of uni-algo. It actually does but only if the CMake option [UNI_ALGO_INSTALL is set](https://github.com/uni-algo/uni-algo/blob/main/CMakeLists.txt#L15). 

How would I handle this best? I would suggest set that option unconditionally in the `generate` method like:
```python
tc.variables["UNI_ALGO_INSTALL"] = True
```

Do you agree?

Reply:
If you add a link to the CMake + use the Conan  2 way for the new generators with `set_property` (we have some docs in the v2 migration page) 

That would be superb 👌

Reply:
@prince-chrismc I've patched. hope I did things right. CMake and how it is modeled in conan has still some mysterious corners for me. 

The CI failed on the commit without any information that helped me to figure out why. If you have access to the backend could you check what's wrong there?

In an blind attempt to fix it I merged master. Let's see if that helps something.